### ISSUE TITLE
Custom Fork command not launching the chosen application

### ISSUE THREAD

**Issue by ImpactKaizer** [sonic2kk/steamtinkerlaunch#563]
Hi ,
been having this issue with STL version 11.0.1 , when every I try any game with it or any proton build it would simply not the load the custom command along side the game I add one game as an example hope it helps.
Feel free to let me know if you require any more details.

[898750.log](https://github.com/frostworx/steamtinkerlaunch/files/9450030/898750.log)

**Comment by sonic2kk** [comment]
Hello! I had a look at the attached log:

```
Tue Aug 30 04:34:15 AM +04 2022 INFO - prepareLaunch - CheckCustomLaunch:
Tue Aug 30 04:34:15 AM +04 2022 INFO - checkCustomLaunch - USECUSTOMCMD is set to '1' - trying to start custom program '/home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe'
Tue Aug 30 04:34:15 AM +04 2022 INFO - checkCustomLaunch - FORK_CUSTOMCMD is set to 1 - forking the custom program in background and continue
Tue Aug 30 04:34:15 AM +04 2022 INFO - launchCustomProg - '/home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' not found - searching in gamedir
Tue Aug 30 04:34:15 AM +04 2022 INFO - launchCustomProg - '/home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' also not in '/home/ahmad/.local/share/Steam/steamapps/common/SRW30//home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' - checking if absolute path was provided
Tue Aug 30 04:34:15 AM +04 2022 INFO - launchCustomProg - '/home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' is absolute path
Tue Aug 30 04:34:15 AM +04 2022 INFO - launchCustomProg - Changed pwd into the custom directory '/home/ahmad/Documents/Trainer'
Tue Aug 30 04:34:15 AM +04 2022 INFO - launchCustomProg - '/home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' seems to be a MS Windows program - starting through proton
Tue Aug 30 04:34:15 AM +04 2022 INFO - launchCustomProg - '/home/ahmad/Documents/Trainer/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' seems to be a MS gui program - starting regularly
```

STL is seeing your custom command and attempting to launch it with the same Proton version that you're running the game with. It seems like somewhere in the background that your custom command is failing.

I'm wondering if perhaps this application you're trying to use just does not work with Wine. I've heard of other people using this Trainer program with STL but I'm not sure if they had to go through any extra steps to get it working. To find out if this is a problem with the Fork option, can you uncheck the Fork option and instead check "Only custom command", and attempt to run your game? And attach a log file too :-) That should help us figure out if this is an STL issue or an issue with your custom command.

STL only attempts to run your custom executable with Proton, but if it doesn't work with Proton/Wine there isn't much that can be done from STL's side.

If you're still having trouble, I remember some discussion around the Trainer program and there needing to be a specific order for game launches for it to work? For example if you need trainer to run before/after a game, take a look at these options:

![image](https://user-images.githubusercontent.com/7917345/187458070-3c29b6b0-4da5-448d-84eb-75c3c451818f.png)

You're currently using Fork Wait, which tries to run your custom program (Trainer) and your game itself. But it will just launch whichever happens to load first. If you need Trainer to load *before* your game, set the "Wait for custom command" value to something. This value (in seconds) is how many seconds to wait after loading the custom executable before launching your game. So if you set this to, say, 10, Trainer would run for 10 seconds and then STL would attempt to run your game.

Likewise, Inject custom command does the opposite. It runs your game then runs your custom command. "Inject wait" is how long you want STL to wait before launching your custom command. So if you needed a game to get through an entire loading sequence or whatever, you could set this to about 15 seconds. That way your game will run for 15 seconds before your custom command is loaded.

I'm not sure if this will help the Trainer program but if you can't get anywhere and you *know* Trainer should work with Wine, that's something to try.

**Comment by sonic2kk** [comment]
Hmm, I don't own Super Robot Wars, but I just tested launching that application with STL and another game and I had no issues. I tested with a git version from a few days ago and a version from the latest git. The program window is a little glitchy on my OS, but it runs. I tested with GE-Proton7-30 though, so you might want to try that version.

Below is my log, which looks pretty much identical to yours minus the paths:

```
Tue Aug 30 15:34:27 BST 2022 INFO - prepareLaunch - CheckCustomLaunch:
Tue Aug 30 15:34:27 BST 2022 INFO - checkCustomLaunch - USECUSTOMCMD is set to '1' - trying to start custom program '/home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe'
Tue Aug 30 15:34:27 BST 2022 INFO - checkCustomLaunch - FORK_CUSTOMCMD is set to 1 - forking the custom program in background and continue
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - '/home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' not found - searching in gamedir
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - '/home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' also not in '/run/media/user/onigiri/Games/steamapps/common/Cookie Clicker//home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' - checking if absolute path was provided
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - '/home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' is absolute path
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - Changed pwd into the custom directory '/home/user/Downloads'
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - '/home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' seems to be a MS Windows program - starting through proton
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - '/home/user/Downloads/Super Robot Wars 30 v1.0-v1.3 Plus 33 Trainer.exe' seems to be a MS gui program - starting regularly
Tue Aug 30 15:34:27 BST 2022 INFO - launchCustomProg - FORK_CUSTOMCMD is set to 1 - forking the custom program in background and continue
```

So I would suggest as well as tinkering with the Inject/Fork wait options, to try a newer GE-Proton version. Though the application worked for me with Proton 7.0-4. The options I mentioned are worth tinkering with in case if/when you get the trainer program working, you may need to set some delays in order to get the application to see the game.

Likely this is an issue specific to the Trainer program running on your system and running through Wine. As the window is glitchy on my system, it's possible that the program is launching but you can't see it because of it being glitchy. You could be able to search for the executable name from your system monitor application, searching for "Trainer" should bring it up. That way you can see if it's running.

I was able to fix the glitchy window on my side by forcing it to have window decorations, which is specific to my desktop environment (KDE Plasma Wayland session on AMD). I could see the window misbehaving on different desktop environments with different GPU drivers and between X11/Wayland. It would probably fix itself if we could run it in a virtual desktop but it doesn't look like there's a way to prepend an argument only for a custom command from STL, so it doesn't look like that's possible right now.

So in short, try a newer GE-Proton version, and check if the trainer program is running at all for you. If you have a second monitor available you could keep your system monitor on that monitor and search for "trainer" before launching the game. Then when you launch the game you can see if it shows up, and if it crashes afterwards.

**Comment by frostworx** [comment]
Those ERROR lines in the log looks "suspicious":
```
Tue Aug 30 04:34:15 AM +04 2022 WARN - extProtonRun - 'proton-unknown-935' seems outdated as the executable /home/ahmad/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --verb=waitforexitandrun -- /home/ahmad/.local/share/Steam/compatibilitytools.d/GE-Proton7-27/proton wasn't found
Tue Aug 30 04:34:15 AM +04 2022 INFO - fixProtonVersionMismatch - Skipping function for 'proton-unknown' Proton version
Tue Aug 30 04:34:15 AM +04 2022 INFO - extProtonRun - Continuing with RUNPROTON='/home/ahmad/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --verb=waitforexitandrun -- /home/ahmad/.local/share/Steam/compatibilitytools.d/GE-Proton7-27/proton'
Tue Aug 30 04:34:15 AM +04 2022 ERROR - extProtonRun - When this error occurs, probably the configured Proton (or a similar) version is no longer available.
Tue Aug 30 04:34:15 AM +04 2022 ERROR - extProtonRun - Because the RUNPROTON '/home/ahmad/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --verb=waitforexitandrun -- /home/ahmad/.local/share/Steam/compatibilitytools.d/GE-Proton7-27/proton' isn't ready here yet - please open an issue when you made sure it is set to an existing one and the error persists
```

Therefore I'd suggest to either check the corresponding line in
`/dev/shm/steamtinkerlaunch/ProtonCSV.txt`
or directly retry with a clean/empty `/dev/shm/steamtinkerlaunch/`
_(and if possible with a clean `~/.config/steamtinkerlaunch/`)_

**Comment by ImpactKaizer** [comment]
Thanks for the reply , I've been using this with the games for a while but after moving to 11.0 from 10 I've noticed this recently as I update once a month (Using Arch) as a reference .
Side note does this have any issues with pipewire by chance ?

Edit: I'm trying with a clean Install of the OS to see that is the case

**Comment by ImpactKaizer** [comment]
Well here's an updated log from a clean install with ProtonGE 7.31
[898750.log](https://github.com/frostworx/steamtinkerlaunch/files/9454305/898750.log)
[ProtonCSV.txt](https://github.com/frostworx/steamtinkerlaunch/files/9454345/ProtonCSV.txt)
Edit : found this intresting on the Beta Channel with Steam UI switch to Stable provides a regular proton / GE name
![image](https://user-images.githubusercontent.com/71709607/187482848-ef78986b-6189-4bd9-b665-2550ba8f1343.png)

**Comment by frostworx** [comment]
thanks. those logs help:
the proton path in ProtonCSV.txt is valid, but in the main log the
`WARN - extProtonRun - 'proton-unknown-929' seems outdated as the executable ` line still shows a broken
proton path
in fact the most relevant line is the line before:
`Tue Aug 30 07:36:58 PM +04 2022 INFO - setRunProtonFromUseProton - Setting ORUNPROTON to '/home/ahmad/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --verb=waitforexitandrun -- /home/ahmad/.local/share/Steam/compatibilitytools.d/GE-Proton7-31/proton'`
and the WARN line is the consequence.
seems like for a still unknown reason the function `setRunProtonFromUseProton` is no longer able to find the correct proton path
_(ORUNPROTON should point to `/home/ahmad/.local/share/Steam/compatibilitytools.d/GE-Proton7-31/proton` and not the whole line it found)_

I've seen you've written version `11.0.1`. That version does not exist, so I assume you mean some git commit version after the last release. Would you please try to reproduce the issue with latest release `11.0`?
Would help a lot to find the regression.

**Comment by ImpactKaizer** [comment]
I'm using the Chaotic AUR build for reference 
But sure I'll install the AUR git build as you requested , also Steam stable build loads it fine while the beta causes this I've noticed

**Comment by sonic2kk** [comment]
I guess for some reason STL is trying to run the game with the Steam Linux Runtime instead of the chosen Proton version, strange.

I'm on my laptop running STL from Git (though not from the AUR), normally I run the latest Steam client beta for my laptop wasn't running it. My desktop is but I do my STL dev stuff on my laptop. I'll investigate and see if I can reproduce the issue, but you might have some option like "Force Steam Linux Runtime" selected in the Game Menu. I would check for that to make sure that isn't checked as it might be ignoring Proton. Also make sure you're running STL as a compatibility tool for this game and not as a launch option.

**Comment by frostworx** [comment]
No, please don't use the AUR git pkgbuild, but the stable one: https://aur.archlinux.org/packages/steamtinkerlaunch
If this can be release nailed down to a Steam version change (I believe you, but maybe we have some false positives or left overs)
this could become problematic, because I almost stopped work on steamtinkerlaunch.

**Comment by sonic2kk** [comment]
Just tried with GE-Proton7-30 and the latest Steam Client Beta, the forked program runs for me with Cookie Clicker at least.

![image](https://user-images.githubusercontent.com/7917345/187487172-d39af164-6f2e-4129-b340-11a7090f7e62.png)

![image](https://user-images.githubusercontent.com/7917345/187487036-4ecb22eb-cd9b-4d6f-870c-6fb69a97f07c.png)

Just to reiterate I don't own Super Robot Wars 30, I'm testing with other games but loading the Trainer program as a forked custom command.

**Comment by ImpactKaizer** [comment]
> Just tried with GE-Proton7-30 and the latest Steam Client Beta, the forked program runs for me with Cookie Clicker at least.

Try it a Fling trainer then & let me know if you could or any Trainer program for that matter unless I need to rename the Trainer to make it short to fix this that would be funny if rename fixes it

**Comment by sonic2kk** [comment]
I believe I'm using a fling trainer (see my edited comment above with a screenshot). Is this the program you're trying to run?

The program window is glitchy for me, popping into and out of existence even when in focus. That's why earlier I suggested checking if the program runs but maybe it isn't showing up for you on your desktop environment. On KDE Plasma I was able to force it to show up by adding window decorations (`Alt+,` by default I think). But this may not be universal.

**Comment by ImpactKaizer** [comment]
Yeah on Stable it loads fine on Beta it starts acting odd but I guess my reply to you got delayed

**Comment by sonic2kk** [comment]
So you're only experiencing problems with the Steam Client Beta? In Steam, could you go to Help > About Steam and take a screenshot of the window that appears?

I'm not so sure this is related to the Steam Beta, but "it works on my machine" (trademark) isn't a great metric to go by :nerd_face: 

Could you check to make sure you aren't trying to force the Steam Linux Runtime?

![image](https://user-images.githubusercontent.com/7917345/187488413-3072f737-bf8b-4514-8f49-6ffc3c5a22ae.png)

**Comment by ImpactKaizer** [comment]
Anyway Here's the Steam Beta with (steamtinkerlaunch **11.0-1**)
I'm adding clean logs for reference as well @frostworx 
![image](https://user-images.githubusercontent.com/71709607/187488547-cdd4b3c3-75c4-47de-bcf1-050987e3bcab.png)
![image](https://user-images.githubusercontent.com/71709607/187488654-c2b414ea-a0f0-42c5-bb69-2001cd40583a.png)
![image](https://user-images.githubusercontent.com/71709607/187488785-1bdc2339-a0e2-481a-b655-e8b1d1fa6bcb.png)
[ProtonCSV.txt](https://github.com/frostworx/steamtinkerlaunch/files/9454661/ProtonCSV.txt)
[898750.log](https://github.com/frostworx/steamtinkerlaunch/files/9454663/898750.log)
![image](https://user-images.githubusercontent.com/71709607/187489077-16106eb4-2234-47b8-9395-e0eed1a4009b.png)

**Comment by ImpactKaizer** [comment]
If you want I can also add logs from the stable build as well to compare if needed

**Comment by sonic2kk** [comment]
This is very strange. Are you definitely running STL as a compatibility tool and not a launch option?

![image](https://user-images.githubusercontent.com/7917345/187491701-5b5bec17-3410-48c8-8c43-2ec44c9874d6.png)

 I'm reading through your updated logs now to see if I can find anything.

**Comment by ImpactKaizer** [comment]
> This is very strange. Are you definitely running STL as a compatibility tool and not a launch option?
> 
> ![image](https://user-images.githubusercontent.com/7917345/187491701-5b5bec17-3410-48c8-8c43-2ec44c9874d6.png)
> 
> I'm reading through your updated logs now to see if I can find anything.

If I install through the AUR it will not load there it's been a thing for a long time there or do I need to add STL to comp.d folder as well ?

**Comment by ImpactKaizer** [comment]
![image](https://user-images.githubusercontent.com/71709607/187492359-f1297837-a55b-44b4-ae7b-09792120917f.png)

**Comment by sonic2kk** [comment]
You probably need to run `steamtinkerlaunch compat add` :smile: and restart Steam after running it. Just like GE-Proton, it won't show up in the list of available compatibility tools until Steam is restarted (this is a Steam thing, it doesn't refresh the list until it restarts).

See the [Installation wiki](https://github.com/frostworx/steamtinkerlaunch/wiki/Installation#package-manager) for details :slightly_smiling_face:

**Comment by ImpactKaizer** [comment]
Cool I'll add it from my side since I mostly choose the ProtonGE & just use `steamtinkerlaunch %command%`

**Comment by sonic2kk** [comment]
The preferred way of launching Proton/Wine games from Steam is to use the Steam Compatibility Tool. Likely the Steam Client Beta has changed how games are launched, maybe it tries to make native Linux stuff (like STL when ran as a launch option?) use the Steam Linux Runtime.

In future it'll probably be more reliable to use your Proton/Wine games as a Compatibility Tool :+1: You can still use SteamTinkerLaunch as a launch option but this is for native Linux titles. See the ["How Do I Use It?" section](https://github.com/frostworx/steamtinkerlaunch#how-do-i-use-it) of the Readme :slightly_smiling_face:

**Comment by frostworx** [comment]
I would have removed the `steamtinkerlaunch %command%` a long time ago already, if linux games worked with steamtinkerlaunch running as compatibility tool...

**Comment by ImpactKaizer** [comment]
Now it's pointing to the to the Trainer fine it seems , Also adding the logs just incase I'm also rebooting just to be sure it loads as well so I wouldn't want to close the STL issue if it persists after reboot 
![image](https://user-images.githubusercontent.com/71709607/187494175-4b1572cf-ad62-4d9e-be48-df19dd83b6cd.png)
[898750.log](https://github.com/frostworx/steamtinkerlaunch/files/9454895/898750.log)
[ProtonCSV.txt](https://github.com/frostworx/steamtinkerlaunch/files/9454903/ProtonCSV.txt)

**Comment by frostworx** [comment]
yay, the weird error is gone \o/
good work @sonic2kk
thanks for good collaboration with debugging, @ImpactKaizer

**Comment by ImpactKaizer** [comment]
> yay, the weird error is gone \o/
> good work @sonic2kk
> thanks for good collaboration with debugging, @ImpactKaizer 

My pleasure I think it's best to remove `steamtinkerlaunch %command%` as it seems it will not play nice with future builds of Steam moving forward.

**Comment by frostworx** [comment]
I'm sure future steam builds _(or friends like SLR)_ will break many things in the future, but until there is no other option to use linux games with steamtinkerlaunch I won't remove it.

**Comment by ImpactKaizer** [comment]
> I'm sure future steam builds _(or friends like SLR)_ will break many things in the future, but until there is no other option to use linux games with steamtinkerlaunch I won't remove it.

Yeah I got the Idea anyway I'm closing this now as we found out the issue with wine related launching thank for the assistance to all
